
To check for potential sample swaps we will compare the 59 SNPs on the EPIC array to genotype data we have from SNP arrays. Genotype data was available for `r sum(!is.na(QCmetrics$genoCheck))` (`r signif(sum(!is.na(QCmetrics$genoCheck))/nrow(QCmetrics),3)*100`%) samples. 

```{r genoCheck, results = "asis", echo = FALSE}

if(class(QCmetrics$genoCheck) == "numeric"){
  hist(QCmetrics$genoCheck, xlab = "Correlation SNP vs DNAm Array", breaks = 50, ylab = "nSamples", main = "")
}

pander(table(QCmetrics$genoCheck > 0.8), caption = "Number of samples where genotypes concordant across SNP and DNAm arrays")
nFail<-sum(QCmetrics$genoCheck < 0.8, na.rm = TRUE)
nRematch<-sum(!is.na(QCmetrics$genoMatch[QCmetrics$genoCheck < 0.8]))


if(nFail > 0){
	write.csv(QCmetrics[which(QCmetrics$genoCheck < 0.8),], file = paste0(dataDir, "/2_gds/QCmetrics/CompSNPdataGeneticErrors.csv"))
}
```

If any of the DNAm samples are not genetically concordant across the EPIC data and SNP chip data then we perform a search to see if they are genetically identical to any other sample in the SNP chip data. This was performed for `r nFail` samples and we identified matching SNP data for `r nRematch` of these individuals. The output of this step has been written to CompSNPdataGeneticErrors.csv.
