
---
title: "QC Stage 1: confirm experimental success"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
params:
  configFile: !r commandArgs(trailingOnly=TRUE)[3]
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(warning = FALSE, message = FALSE, fig.align='center', echo = FALSE)

# load packages
library(dplyr)
library(plyr)
library(knitr)
library(vioplot)
library(corrplot)
library(ggplot2)
library(reshape2)
library(gridExtra)
library(cowplot)
library(kableExtra)
library(scales)
library(tibble)
library(ChIPQC)
library(plotly)
options(scipen=5)
## load config variables
source(params$configFile)

```

This report provides a summary of an ATAC-seq experiment from the resulting sequencing data. It looks at metrics from the raw sequencing data, alignment and filtering (and peak calling steps). The bioinformatic pipeline parallels that of the [ENCODE ATAC-seq guidelines](https://www.encodeproject.org/atac-seq/). Raw fastq files have been have been aligned with Bowtie2 to the human genome (hg38) and peaks called with MACS3 for each individual sample. The pipeline is still under development.

```{r loadDat, message = FALSE, results='hide'}
## import data
pheno<-read.table(sampleSheet, header = TRUE, sep = ',', stringsAsFactors = FALSE)
processSum <- read.csv(file.path(metaDir, "/summariseSampleProcessingProgress.csv"), stringsAsFactors = FALSE, strip.white = TRUE)
fastqc<-read.table(file.path(fastQCDir, "/multiqc/multiqc_data/multiqc_fastqc.txt"), sep = "\t", header = TRUE, stringsAsFactors = FALSE)
readCounts<-read.table(file.path(alignedDir, "/countMTReads.txt"), fill = TRUE, skip = 1)
alignQC<-read.table(file.path(alignedDir, "/multiqc/multiqc_data/multiqc_bowtie2.txt"), sep = "\t", header = TRUE, stringsAsFactors = FALSE)
flagstat<-read.csv(file.path(alignedDir, "/ENCODEMetrics/collateFlagStatMetrics.txt"), header = FALSE)
flagstat<- flagstat[,c(1:ncol(flagstat)-1)]

# wrangle data
missingCol<-setdiff(c("sequencingBatch", "sampleID",  "cohort", "fraction", "experiment", "individualID"), colnames(pheno))
if (length(missingCol) != 0 ) {
  if ("sequencingBatch" %in% missingCol){
    pheno$sequencingBatch<-pheno$cohort
      if (length(missingCol) > 1) {
      warning('Incorrect column names. Missing column(s): ', list(missingCol))
      stop()
    }
  }
}
## exclude duplicates from metadata
dups<-names(which(table(paste(pheno$individualID, pheno$fraction, pheno$tissue, sep = "_")) > 1))
pheno<-unique(pheno[,intersect(c("sequencingBatch", "sampleID","cohort","fraction","individualID", "tissue"), colnames(pheno))])
processSum <- processSum[match(pheno$sampleID, processSum$sampleID),]

## wrangle multiqc stats to retain important columns
colsKeep<-c("Total.Sequences","total_deduplicated_percentage", "Sequences.flagged.as.poor.quality")
mergeStats<-cbind(fastqc[match(processSum$R1Filename, fastqc$Filename),colsKeep], fastqc[match(processSum$R2Filename, fastqc$Filename),colsKeep])

## calculate percentage of mitochondrial reads
readCounts<-readCounts[match(pheno$sampleID, readCounts$V1),]
mergeStats<-cbind(mergeStats, readCounts$V2/readCounts$V3*100)
colnames(mergeStats)[ncol(mergeStats)]<-"PercentMTReads"

## select alignment columns and match to sampleIDs
aIndex<-match(processSum$sampleID, gsub("\\.bowtie", "", alignQC$Sample))
alignQC$aligned_total<-rowSums(alignQC[,c("paired_aligned_one", "paired_aligned_multi")])
mergeStats<-cbind(mergeStats,alignQC[aIndex,c("overall_alignment_rate", "aligned_total")])
mergeStats<-cbind(sample=processSum$sampleID, mergeStats)

## load flagstat metrics calculated as part of encode qc pipeline
flagstat<-flagstat[match(processSum$sampleID, flagstat$V1),]
files<-list.files(paste0(alignedDir, "/ENCODEMetrics"), pattern = ".pbc.qc")
eMetrics<-NULL
for(each in files){
  tmp<-read.table(paste0(alignedDir, "/ENCODEMetrics/", each))
  if(ncol(tmp) == 7){
    eMetrics<-rbind(eMetrics, tmp)
  } else {
    eMetrics<-rbind(eMetrics, rep(NA, 7))
  }
}
colnames(eMetrics)<-c("TotalReadPairs","DistinctReadPairs","OneReadPair","TwoReadPairs","NRF","PBC1","PBC2")
eMetrics$Sample <- files
eMetrics$Sample <- gsub(".pbc.qc", "", eMetrics$Sample)
eIndex<-match(processSum$sampleID, gsub(".pbc.qc", "", files))
eMetrics<-eMetrics[eIndex,]
mergeStats<-cbind(mergeStats,eMetrics)

## Option to use a single csv file with all results
fripStats <- read.csv(file.path(peakDir, "/QCOutput/FRIP_all_samples.csv"), stringsAsFactors = FALSE, strip.white = TRUE)
fripStats$FripMACS3PE <- fripStats$ReadsinBAMPeaks/fripStats$BAMTotalReads
fripStats[is.na(fripStats)] <- 0
fripStats <- fripStats[match(pheno$sampleID, fripStats$SampleName),]

#ccFiles<-list.files(alignedDir, pattern = '.subsample.cc.qc')
#crossCor<-read.csv(paste(alignedDir, ccFiles[1], sep='/'))
#for(each in ccFiles[-1]){
#  crossCor<-rbind(crossCor, read.csv(paste0(alignedDir, each,  sep='/')))
#}
#crossCor<-crossCor[match(processSum$sampleID, crossCor$SampleName),]

## load periodicity metrics 
propNucleosomesAll<-NULL
diptestStatsAll<-NULL
filePaths <- list.files(file.path(alignedDir,"/QCOutput/"), pattern = "FragmentDistribution_Batch|FragmentDistribution_Batch_")
for(each in filePaths){
  load(paste0(alignedDir,"/QCOutput/", each))
  propNucleosomesAll<-rbind(propNucleosomesAll, propNucleosomes)
  diptestStatsAll<-rbind(diptestStatsAll, diptestStats)
}

#diptestStatsAll<-as.data.frame(diptestStatsAll)
propNucleosomesAll<-propNucleosomesAll[match(processSum$sampleID, rownames(propNucleosomesAll)),]
#diptestStatsAll<-diptestStatsAll[match(processSum$sampleID, sub("X", "", gsub("\\.D$", "", rownames(diptestStatsAll)))),]
diptestStatsAll<-diptestStatsAll[match(processSum$sampleID, gsub("\\.D$", "", rownames(diptestStatsAll))),]
rownames(diptestStatsAll)<-processSum$sampleID

## count number of samples with X million reads
readThres<-seq(0,max(mergeStats[,2], na.rm = TRUE)+10^6, 10^6)
nSamples<-matrix(data = NA, nrow = length(readThres), ncol = length(unique(pheno$sequencingBatch))+1)
for(i in readThres){
  nSamples[1+(i/10^6),1] <- sum(mergeStats[,2] > i, na.rm = TRUE)
  colNum<-2
  for(each in unique(pheno$sequencingBatch)){
    nSamples[1+(i/10^6),colNum] <- sum(mergeStats[which(pheno$sequencingBatch == each),2] > i, na.rm = TRUE)
    colNum<-colNum+1
  }
}

#propNucleosomesAll <- na.omit(propNucleosomesAll)
## a successful ATAC experiment should have decreasing proportions in the nucleosome, mono, di,tri- nucleosomes
## define a tolerance for these comparisons
tol <- 0
decreasingProps<-cbind(propNucleosomesAll[,1]+tol > propNucleosomesAll[,2], 
                       propNucleosomesAll[,2]+tol > propNucleosomesAll[,3], 
                       propNucleosomesAll[,3]+tol > propNucleosomesAll[,4]
)
colnames(decreasingProps)<-c("NFR>Mono", "Mono>Di", "Di>Tri")
decreasingProps <- as.data.frame(decreasingProps)
```


Data was loaded for `r nrow(pheno)` samples from `r length(unique(pheno$individualID))` individuals. The number by sample type is shown below.


```{r, echo = FALSE}
nCTStart<-table(pheno$fraction)

kbl(nCTStart) %>%
  kable_styling(bootstrap_options = c("striped", "hover"),font_size = 10)

```


## Overview

To provide an overall summary here is a plot of the number of raw reads, aligned reads (post filtering) and peaks called for each sample.

<style>
  .scrollingimage{
      overflow-x:scroll;
      white-space: nowrap;
  }

  .scrollingimage img{
     max-width: none;
  }

</style>
```{r setup2}
if (nrow(pheno) > 70) {
  ## include scrolling bar if number of samples exceeds 70
  knitr::opts_chunk$set(fig.width=(nrow(pheno)/5))
}
```

<div class="scrollingimage">
```{r overview, echo=FALSE}
# plot barchart of 
mergeStats <- mergeStats[,-c(1,17)]
data.frame(pheno$sampleID, mergeStats[,c(1,9)],fripStats$ReadsinBAMPeaks) %>%
  setNames(c('Samples', 'Raw reads', 'Aligned reads', 'Reads in peaks')) %>%
  melt(id='Samples') %>%
  ggplot(., aes(fill= variable, x=Samples, y=value))+
    geom_bar(position = 'dodge', stat = 'identity')+
    theme_bw()+
    labs(y = 'Number of reads')+
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          legend.position = 'bottom', 
          legend.title = element_blank())+
    scale_fill_manual(values = colorBlindGrey8)

knitr::opts_chunk$set(fig.width=7)
```
</div>

## 1. Sequencing

### 1.1 Data summary

In total there were `r signif(sum(c(mergeStats[,1], mergeStats[,4]), na.rm = TRUE)/1000000,3)` million reads generated across all samples, with a mean of 
`r signif(mean(c(mergeStats[,1], mergeStats[,4]), na.rm = TRUE)/1000000,3)` million per sample (SD = `r signif(sd(c(mergeStats[,1], mergeStats[,4]), na.rm = TRUE)/1000000,3)` million). 

The median number of reads was  `r signif(median(c(mergeStats[,1], mergeStats[,4]), na.rm = TRUE)/1000000,3)` million per sample (IQR = `r signif(sd(c(mergeStats[,1], mergeStats[,4]), na.rm = TRUE)/1000000,3)` million).

The plots below summarise the distributions of the raw sequencing data for `r sum(!is.na(mergeStats[,1])) ` samples. For mammalian studies, authors of the ATAC-seq protocol recommend a minimum of 50 million non-duplicated, non-mitochondrial mapped reads for open chromatin detection. The ENCODE pipeline recommends a minimum of 25 million for single end and 50 million for paired. In the plots below, the median of the data is indicated by the orange line and 50 and 25 million thresholds by the black lines.

```{r pressure, fig.width = 10, fig.height = 12}
# plot histograms of read numbers and % unique reads
xlab <- c('Number of Reads', '% unique reads' ,'' , 'Number of Reads', '% unique reads')
read<-c('R1','R1','','R2', 'R2')
p=NULL
for (x in 1:5){
  p[[x]]<-mergeStats %>% dplyr::select(all_of(x))%>%
          ggplot(aes(x = .[,1]))+ # select x values based on column index rather than name
          geom_histogram(bins=100, fill ='white', colour = 'black')+
          labs(title = read[x], x = xlab[x], y = 'Number of Samples')+
          theme_bw() 
}
p<-p[- 3] # remove the empty list item
## add thresholds
p[[1]]<-p[[1]]+ geom_vline(xintercept = median(mergeStats[,1], na.rm = TRUE), color = "orange") +
  geom_vline(xintercept = c(25000000,50000000), color = "black") +
	scale_x_continuous(labels = label_number(suffix = "M", scale = 1e-6))
p[[3]]<-p[[3]]+ geom_vline(xintercept = median(mergeStats[,1], na.rm = TRUE), color = "orange") +
  geom_vline(xintercept = c(25000000,50000000), color = "black")+
	scale_x_continuous(labels = label_number(suffix = "M", scale = 1e-6))

grid.arrange(grobs = p,ncol=2)
```

The plot below presents the same data in a slightly different way, counting the number of samples that have that quantity of data. Each line represents a different sequencing batch.

``` {r minReads}
# plot minimum number of reads
coeff=nSamples[1,1]/100
data.frame(readThres/10^6, nSamples[,-1]) %>% 
  melt(., id='readThres.10.6') %>%
  ggplot(aes(x = .[,1], y = .[,3], colour = .[,2]))+
  geom_line() +
  labs(x = 'Minimum number of reads (millions)', y = 'Number of Samples')+
  theme_bw()+
  geom_vline(xintercept=c((25)/0.8, 50, median(mergeStats[,1], na.rm = TRUE)/10^6), 
             color = c('black', 'black', colorBlindGrey8[2]),
             alpha=0.9, linetype="dashed", show.legend =FALSE)+
  scale_y_continuous(sec.axis = sec_axis(transform = ~./coeff, name = '% of samples'))+
  theme(legend.position = "none")+
  scale_color_manual(values = rep('black', ncol(nSamples)))
```


A good ATAC-seq experiment will capture enrichment at a broad range of genomic locations. The number of distinct or unique reads is an indicator of the diversity or complexity of the data. If the sample does not contain sufficient DNA and/or has been over-sequenced, you will simply be repeatedly sequencing PCR duplicates of a restricted pool of distinct DNA fragments. This is known a low-complexity library and is not desirable.  


``` {r uniqueReads}
## plot %unique reads against read number
p <- ggplot(mergeStats[,c(1,2)], aes(x = Total.Sequences, y = total_deduplicated_percentage, colour = pheno$fraction, label= mergeStats$Sample))+
  geom_point() +
  labs(x = 'Number of Reads', y = '% unique reads')+
  theme_bw() +
  scale_color_manual(values = colorBlindGrey8[1:length(unique(pheno$fraction))])+
  theme(legend.title = element_blank(),
        legend.position = 'bottom')
ggplotly(p, tooltip = "label")
```

``` {r violinUniqueReads, fig.width = 10, fig.height = 15}
## plot no reads and %unique by sequencing batch
p<-NULL
x=1
cols<-c("sequencingBatch", "fraction", "cohort", "Sequencing batch", "Cell fraction", "Brain bank")
lab <- c('Number of reads', '% unique reads', "Alignment rate", "Number of aligned reads")
for (i in 1:3){ #select pheno columns
  for (j in 1:2){ #select mergeStats columns
    p[[x]]<-data.frame(mergeStats[,j], pheno[,cols[i]]) %>%
      ggplot(aes(y = .[,1], x=as.factor(.[,2]), fill = as.factor(.[,2])))+
      geom_violin(scale = 'width')+
      geom_boxplot(width=0.05, fill='black')+
      labs(x = cols[i+3], y = lab[j])+
      stat_summary(fun="median", 
                 geom="point", color="white")+
      theme_bw() +
      theme(axis.text.y = element_text(angle = 90), 
            axis.text = element_text(size = 7),
            axis.title = element_text(size=10),
            legend.position="none")
    x<-x+1
  }
}
grid.arrange(grobs = p,ncol=2)

```

## 2. Alignment


### 2.1 Summary of alignment (pre-filtering)

The plot below shows the alignment rate and number of mapped reads per sample. While general convention is a unique mapping rate over 80% is typical for a successful ATAC-seq experiment, ENCODE would require that the percentage of mapped reads should be greater than 95%, though values >80% may be acceptable. In the plot below the pink line indicates the 80% threshold, the black indicates the 95%. The plot below contains data for `r sum(!is.na(mergeStats$aligned_total))` samples. 

```{r alignment rate}
## plot alignment rate against number of reads
p <-ggplot(mergeStats[,c('aligned_total','overall_alignment_rate',"Sample")], aes(x = aligned_total, y = overall_alignment_rate, colour = pheno$fraction, label = Sample))+
  geom_point() +
  geom_hline(yintercept=c(95, 80), #acceptable
             color = c('black', colorBlindGrey8[2]))+
  labs(x = 'Number of Reads', y = 'Alignment rate (%)')+
  theme_bw() +
  scale_color_manual(values = colorBlindGrey8[1:length(unique(pheno$fraction))])+
  theme(legend.title = element_blank(),
        legend.position = 'bottom')+
	scale_x_continuous(labels = label_number(suffix = "M", scale = 1e-6))
ggplotly(p, tooltip = "label")
```


``` {r alignBoxplot, fig.width = 10, fig.height = 12}
## plot alignment rate and no aligned reads by sequencing batch, cell fraction and brain bank
p<-NULL
x=1
for (i in 1:3){
  for (j in 8:9){
    p[[x]]<-data.frame(mergeStats[,j], pheno[,cols[i]]) %>%
      ggplot(aes(y = .[,1], x=as.factor(.[,2]), fill = as.factor(.[,2])))+
      geom_violin(scale='width')+
      geom_boxplot(width=0.05, fill='black')+
      labs(x = cols[i+3], y = lab[j-5])+
      stat_summary(fun="median", 
                 geom="point", color="white")+
      theme_bw()+
      theme(axis.text.y = element_text(angle = 90), 
            axis.text = element_text(size = 7),
            axis.title = element_text(size=10), legend.position = "none")
    
    if (j==8){
      p[[x]]<-p[[x]]+geom_hline(yintercept = 80, alpha=0.7, linetype="dashed", show.legend =FALSE)
    }
    x<-x+1
  }
}
grid.arrange(grobs = p,ncol=2)

```


### 2.2 Mitochondrial genome

It can be a useful QC to assess the number of mitochondrial reads; some ATAC-Seq samples have been reported to be 80% mitochondrial reads. 

``` {r mitochondria}
# plot percentage of mitochondrial reads
data.frame(mergeStats$PercentMTReads, pheno$cohort) %>%
  ggplot(aes(y = .[,1], x=as.factor(.[,2]), fill = as.factor(.[,2])))+
      geom_violin(scale='width')+
      geom_boxplot(width=0.05, fill='black')+
      labs(x = "Brain bank", y = "Percentage of MT reads")+
      stat_summary(fun="median", 
                 geom="point", color="white")+
      theme_bw()+
  theme(legend.position = "none")

data.frame(mergeStats$PercentMTReads, pheno$sequencingBatch) %>%
  ggplot(aes(y = .[,1], x=as.factor(.[,2]), fill = as.factor(.[,2])))+
      geom_violin(scale='width')+
      geom_boxplot(width=0.05, fill='black')+
      labs(x = "Sequencing batch", y = "Percentage of MT reads")+
      stat_summary(fun="median", 
                 geom="point", color="white")+
      theme_bw()+
  theme(legend.position = "none")


```

### 2.3 Library Complexity

In this section we will assess the number of unique aligned reads as the number of total reads increases, across all samples. 

``` {r dedupreads}
# plot unique reads against total reads
libcomp <- data.frame(x = mergeStats[,1], y = mergeStats$DistinctReadPairs, fraction = pheno$fraction, name =pheno$sampleID) 
p <- ggplot(libcomp,aes(x=x, y = y, colour = fraction,label = name))+
      geom_point()+
      theme_bw()+
      labs(x = "Number of reads", y = "Number of distinct aligned reads")+
      scale_color_manual(values = colorBlindGrey8[1:length(unique(pheno$fraction))])+
      theme(legend.title = element_blank(), 
            legend.position = 'bottom') +
	scale_x_continuous(labels = label_number(suffix = "M", scale = 1e-6))+
	scale_y_continuous(labels = label_number(suffix = "M", scale = 1e-6))
ggplotly(p, tooltip = "label")
```

We will also calculate additional statistics recommended by the ENCODE ATAC-seq pipeline: PCR Bottlenecking Coefficients 1 & 2 (PBC1 & PBC2) and Non-Redundant Fraction (NRF) and compare these to their reference values. The plots below contain data from `r sum(!is.na(eMetrics[,1]))` samples.

``` {r encode, fig.height = 12}
## plot encode statistics
p=NULL
## to ensure the threshold labels are in correct place according to threshold lines, calculate midpoints between lines:
# matrix of encode thresholds
thres<-matrix(c(0.7, 0.9,
               0.7, 0.9, 
               1, 3), nrow = 2)
# matrix of label orientation (mean)
labs<-matrix(data=NA, ncol = 3, nrow = 3)
labs[2, ]<-colMeans(thres)
labsText<-list(c("Concerning", "Acceptable", "Ideal"), c("Severe", "Concerning", "None"), c("Severe", "Concerning", "None"))

for (x in 1:3){
  colNum<-x+4 # index to correct column according to label
  min<-min(eMetrics[,colNum], na.rm = TRUE)
  max<-max(eMetrics[,colNum], na.rm = TRUE)
  # generate the placement of the labels relative to the data
  if (min < thres[1, x]){
    labs[1, x]<-mean(c(min, thres[1, x]))
  } else { labs[1, x]<-thres[1, x]-sd(eMetrics[,colNum], na.rm = TRUE) } 
  if(max > thres[2, x]){
    labs[3, x]<-mean(c(max, thres[2, x]))
  } else { labs[3, x]<- thres[2, x]+sd(eMetrics[,colNum], na.rm = TRUE) }
  
  # create plot
  p[[x]]<-dplyr::select(eMetrics, all_of(colNum)) %>%
    ggplot(aes_string(x=names(.)[1]))+
    geom_histogram(bins = 20, fill ='white', colour = 'black')+
    labs(y = "Number of samples")+
    geom_vline(xintercept = c(thres[,x]))+
    annotate("text", x = labs[,x], y = Inf, vjust = 2, label = labsText[[x]])+
    theme_bw()+
    theme(axis.text.y = element_text(angle = 90))
  
}
p<-compact(p)
grid.arrange(grobs = p, nrow = 3)

```


### 2.4 Fragment Size Distribution

ATAC-seq data should be enriched for specific sized fragments, so the distribution of fragment size should show periodicity. Typically there should be a large proportion of reads with fragment length < 100bp indicating nucleosome-free fragments, the plot should then show periodicity of ~ 200bp representing  enrichment of mono-, di-, and tri-nucleosome bound fragments (~200, 400, 600bp, respectively). Typically we would look at the fragment distribution where a successful ATAC-seq experiment would have a peak at ~50-100bp with a large proportion of reads less than 100 bp, and peaks at 200bp, 400bp, 600bp. If there was no chromatin in the sample the graph would have no periodicity and resemble a poisson distribution with lambda = 1). Instead to make this tractable across hundred of samples we have classified the fragments into bins representing nucleosome-free,mono-, di-, tri, and quad-nucleosome bound fragments. From this for each sample we can calculate the proportion within in each bin and apply filters/conditions. If in general the ATAC-seq has worked, we would expect to see decreasing proportions across the violins below. The plot below contains data from `r sum(!is.na(propNucleosomesAll[,1]))` samples.  

``` {r periodicity}
## plot nucleosome periodicity
melt(na.omit(propNucleosomesAll)) %>%
  ggplot(aes(x=as.factor(Var2), y=value))+
      geom_violin(fill = "grey80", scale = 'width')+
      geom_boxplot(width=0.05, fill='black')+
      labs(y = "Proportion of reads", x = "Nucleosome")+
      stat_summary(fun="median", geom="point", color="white")+
      scale_x_discrete(labels = c("NFR", "Mono", "Di", "Tri", "Quad"))+
      theme_bw()
```

We need to apply thresholds to satisfy the following criteria

* large proportion of nucleosome-free fragments
* presence of mono-nucleosome bound fragments
* decreasing number of fragments in fragment bins.

```{r periodicity2}

# plot proportion of nucleosome free against mono
prop <- data.frame(propNucleosomesAll)
p <-ggplot(prop, aes(x=propNucleosomesAll[,2], y=propNucleosomesAll[,1]))+
    geom_point()+
    theme_bw()+
    labs(x= "Proportion fragments classed as mono", y = "Proportion fragments class as nucleosome free")+
    geom_abline(intercept = 0, slope = 1)+
    ylim(c(0,1))+
    xlim(c(0,max(propNucleosomesAll[,2])))+
    geom_hline(yintercept = c(nucfThres))+
    geom_vline(xintercept = c(monoThres)) +
    geom_polygon(data = data.frame("x" = c(monoThres, nucfThres, 1,monoThres), "y" = c(nucfThres, nucfThres, 1, 1)), aes(x = x, y = y), fill = "grey", alpha = 0.2, size = 2)

p
```

Alternatively we can test for periodicity (i.e. multimodality) using Hartigan's Dip test for Unimodality. Here a significant p-value indicates the presence of multiple modes as we would expect in a successful ATAC experiment. The plot below contains data from `r sum(!is.na(diptestStatsAll[,1]))` samples.  


```{r periodicity3}

# plot distribution of hartigan's dip test p values
hist(diptestStatsAll[,"p.value"], breaks = 50, col = "grey", xlab = "Hartigans Test P-value", ylab = "Number of samples", main = "")
abline(v = dipPThres)

```

## 3. Peak calling

### 3.1 Number of peaks

Peak calling was performed in three ways using MACS3 in the paired end mode. ENCODE recommends that number of peaks in each sample be greater than 150,000, though values greater than 100,000 are acceptable. Data from `r sum(!is.na(fripStats$BAMPeaks))` samples is plotted below. The mean number of peaks for MACS3 PE Peaks is `r signif(mean(fripStats$BAMPeaks, na.rm = TRUE ),3)` (SD = `r signif(sd(fripStats$BAMPeaks, na.rm = TRUE ),3)`). The plots below contain data from `r sum(!is.na(fripStats$BAMPeaks))` samples. 

```{r numpeaks, results=FALSE}
data.frame(fripStats$BAMPeaks)%>%
  melt() %>%
  ggplot(aes(y = value, x=variable))+
      geom_violin(fill = "grey80", scale='width')+
      geom_boxplot(width=0.05, fill='black')+
      labs(x = "Sequencing batch", y = "Number of peaks (100000s)")+
      stat_summary(fun="median", 
                 geom="point", color="white")+
      geom_hline(yintercept = c(150000, 100000), color = c('black', colorBlindGrey8[3]))+
      scale_x_discrete(labels = "MACS3 Paired-end")+
	scale_y_continuous(labels = label_number(suffix = "", scale = 1e-5))+
      theme_bw()

```

```{r rviolinPlotPeaks,fig.height = 8, fig.width = 8}
p <- NULL
p[[1]]<-data.frame(fripStats$BAMPeaks, pheno$cohort) %>%
      ggplot(aes(y = .[,1], x=as.factor(.[,2]), fill = as.factor(.[,2])))+
      geom_violin(scale='width')+
      geom_boxplot(width=0.05, fill='black')+
      labs(x = "Brain bank", y = "Number of peaks PE (100000s)")+
      stat_summary(fun="median", 
                 geom="point", color="white")+
      theme_bw()+
      theme(axis.text.y = element_text(angle = 90), 
            axis.text = element_text(size = 7),
            axis.title = element_text(size=10), legend.position = "none")+
	scale_y_continuous(labels = label_number(suffix = "", scale = 1e-5))

p[[2]]<-data.frame(fripStats$BAMPeaks, pheno$fraction) %>%
      ggplot(aes(y = .[,1], x=as.factor(.[,2]), fill = as.factor(.[,2])))+
      geom_violin(scale='width')+
      geom_boxplot(width=0.05, fill='black')+
      labs(x = "Cell fraction", y = "Number of peaks PE (100000s)")+
      stat_summary(fun="median", 
                 geom="point", color="white")+
      theme_bw()+
      theme(axis.text.y = element_text(angle = 90), 
            axis.text = element_text(size = 7),
            axis.title = element_text(size=10), legend.position = "none")+
	scale_y_continuous(labels = label_number(suffix = "", scale = 1e-5))

p[[3]]<-data.frame(fripStats$BAMPeaks, pheno$sequencingBatch) %>%
      ggplot(aes(y = .[,1], x=as.factor(.[,2]), fill = as.factor(.[,2])))+
      geom_violin(scale='width')+
      geom_boxplot(width=0.05, fill='black')+
      labs(x = "Sequencing Batch", y = "Number of peaks PE (100000s)")+
      stat_summary(fun="median", 
                 geom="point", color="white")+
      theme_bw()+
      theme(axis.text.y = element_text(angle = 90), 
            axis.text = element_text(size = 7),
            axis.title = element_text(size=10), legend.position = "none")+
	scale_y_continuous(labels = label_number(suffix = "", scale = 1e-5))

grid.arrange(grobs = p,ncol=2)
```

Peak calling will be affected by the background noise of a sample, and the success of the ATAC enrichment. Let's compare some of our periodicity statistics against the number of peaks. 

Below, results from calling peaks using MACS3 PE mode are shown. 
```{r, fig.height = 16 , fig.width = 8}
p<-NULL
p[[1]]<-data.frame(propNucleosomesAll)%>%
  ggplot(aes(x=fripStats$BAMPeaks, y=propNucleosomesAll[,1]))+
    geom_point()+
    theme_bw()+
    labs(x= "Number of Peaks (x10000)", y = "Proportion nucleosome free")+
	scale_x_continuous(labels = label_number(suffix = "", scale = 1e-5))

p[[2]]<-data.frame(propNucleosomesAll)%>%
  ggplot(aes(x=fripStats$BAMPeaks, y=propNucleosomesAll[,2]))+
    geom_point()+
    theme_bw()+
    labs(x= "Number of Peaks (x10000)", y = "Proportion mono-nucleosome") +
	scale_x_continuous(labels = label_number(suffix = "", scale = 1e-5))

p[[3]]<-data.frame(diptestStatsAll)%>% 
    ggplot(aes(x=fripStats$BAMPeaks, y=diptestStatsAll[,"D"]))+
    geom_point()+
    theme_bw()+
    labs(x= "Number of Peaks (x10000)", y = "Hartigans Test Statistic") +
	scale_x_continuous(labels = label_number(suffix = "", scale = 1e-5))

p[[4]]<-data.frame(diptestStatsAll)%>% 
    ggplot(aes(x=fripStats$BAMPeaks, y=diptestStatsAll[,"p.value"]))+
    geom_point()+
    theme_bw()+
    labs(x= "Number of Peaks (x10000)", y = "Hartigans Test P") +
	scale_x_continuous(labels = label_number(suffix = "", scale = 1e-5))


p[[5]]<-data.frame(propNucleosomesAll)%>%
  ggplot(aes(x=fripStats$FripMACS3PE, y=propNucleosomesAll[,1]))+
    geom_point()+
    theme_bw()+
    labs(x= "FRIP", y = "Proportion nucleosome free")

p[[6]]<-data.frame(propNucleosomesAll)%>%
  ggplot(aes(x=fripStats$FripMACS3PE, y=propNucleosomesAll[,2]))+
    geom_point()+
    theme_bw()+
    labs(x= "FRIP", y = "Proportion mono-nucleosome")


p[[7]]<-data.frame(diptestStatsAll)%>% 
    ggplot(aes(x=fripStats$FripMACS3PE, y=diptestStatsAll[,"D"]))+
    geom_point()+
    theme_bw()+
    labs(x= "FRIP", y = "Hartigans Test Statistic")

p[[8]]<-data.frame(diptestStatsAll)%>% 
    ggplot(aes(x=fripStats$FripMACS3PE, y=diptestStatsAll[,"p.value"]))+
    geom_point()+
    theme_bw()+
    labs(x= "FRIP", y = "Hartigans Test P")


grid.arrange(grobs = p, ncol=2, nrow = 4)

```

### 3.2 Specificity of reads in peaks

Plot of the fraction of reads in peaks called for that sample. These plots give a good indication of how "enriched" the sample is and can be considered a "signal-to-noise" measure of what proportion of the library consists of fragments from binding sites vs. background reads. According to the ENCODE standard processing pipeline for ATAC-seq the FRiP score should be > 0.3, though values greater than 0.2 are acceptable. We will not filter on this metric yet, but return to it once we have called peaks at the group level. 

```{r readsinpeaks}
# plot 
data.frame(fripStats$FripMACS3PE)%>%
  melt() %>%
  ggplot(aes(y = value, x=variable))+
      geom_violin(fill = "grey80", scale='width')+
      geom_boxplot(width=0.05, fill='black')+
      labs(x = "Sequencing batch", y = "Fraction reads in peaks")+
      stat_summary(fun="median", 
                 geom="point", color="white")+
      geom_hline(yintercept = c(0.3, 0.2), color = c('black', colorBlindGrey8[3]))+
      scale_x_discrete(labels = "MACS3 Paired-end")+
      theme_bw()


p<- ggplot(fripStats, aes(x=BAMPeaks, y = FripMACS3PE, colour = pheno$fraction, label = SampleName))+
    geom_point()+
    theme_bw()+
    labs(title = 'MACS3: PE', x = "Number of peaks (100000s)", y = "Fraction reads in peaks")+
    scale_color_manual(values = colorBlindGrey8[1:length(unique(pheno$fraction))])+
    scale_x_continuous(labels = label_number(suffix = "", scale = 1e-5))+
    theme(legend.title = element_blank(), 
          legend.position = 'bottom')
ggplotly(p, tooltip ="label")
```

# QC Summary: Sample Filtering

```{r samplePASS}
# generate a txt file of passed samples
QCPASS<-cbind(mergeStats$overall_alignment_rate > alignThres, #alignment rate 
              fripStats$BAMTotalReads > uniqReads*10^6, #total filtered & aligned reads
              eMetrics$NRF > nrfThres, #encode
              eMetrics$PBC1 > pbc1Thres, #encode
              eMetrics$PBC2 > pbc2Thres, #encode,
              propNucleosomesAll[,1] > nucfThres, #fractions nucleosome free 
              propNucleosomesAll[,2] > monoThres, #fractions mononucleosome 
              decreasingProps[,1],#periodicity
              fripStats$BAMPeaks > peakThres, # number of peaks
              fripStats$FripMACS3PE > fripThres, #frip
              diptestStatsAll[,"p.value"] < dipPThres #multimodality
              ) 
QCPASS[is.na(QCPASS)] <- FALSE
colnames(QCPASS)<-c("AlignmentRate", "UniqueReads", "NRF", "PBC1", "PBC2","NFR", "MNC", "Periodicity","Peaks", "FRIP", "Multimodality")
rownames(QCPASS)<-processSum$sampleID
keep<-rowSums(QCPASS, na.rm=TRUE) == rowSums(!is.na(QCPASS))
QCPASS<-cbind(QCPASS, "QCPASS" = keep)
QCPASS.df<- cbind(pheno, QCPASS)
write.csv(QCPASS.df, file = file.path(metaDir, "/passStage1Status.csv"))

write.table(names(keep[keep==TRUE]), file = file.path(metaDir, "/passStage1SampleList.txt"), 
            row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(pheno1, file = file.path(metaDir, "/matchedVCFIDs.txt"), 
            row.names = FALSE, col.names = FALSE, quote = FALSE)


qcStats <- cbind(pheno, mergeStats$overall_alignment_rate, #alignment rate 
              fripStats$BAMTotalReads, #total filtered & aligned reads
              eMetrics$NRF, #encode
              eMetrics$PBC1, #encode
              eMetrics$PBC2, #encode,
              propNucleosomesAll[,1], #fractions nucleosome free 
              propNucleosomesAll[,2], #fractions mononucleosome 
              decreasingProps[,1],#periodicity
              fripStats$BAMPeaks, # number of peaks
              fripStats$FripMACS3PE, #frip
              diptestStatsAll[,"p.value"] #multimodality
              ) 

colnames(qcStats) <- c(colnames(pheno), "AlignmentRate", "UniqueReads", "NRF", "PBC1", "PBC2","NFR", "MNC", "Periodicity","Peaks", "FRIP", "Multimodality")
write.csv(qcStats, file = file.path(metaDir, "/qc1Stats.csv"),row.names = FALSE)
```

Samples will now be filtered based on thresholds defined in the config file. The table below, summarises the number of samples that fail each QC step.

```{r, echo = FALSE}

kbl(colSums(QCPASS, na.rm = TRUE))

# create cross tab table
xTab<-matrix(data = NA, ncol = ncol(QCPASS), nrow = ncol(QCPASS))
for(i in 1:ncol(QCPASS)){
  for(j in 1:ncol(QCPASS)){
    xTab[i,j]<-sum(rowSums(QCPASS[,c(i,j)]) == 0, na.rm = TRUE)
  }
}
colnames(xTab)<-colnames(QCPASS)
rownames(xTab)<-colnames(QCPASS)

kbl(xTab) %>%
  kable_styling(bootstrap_options = c("striped", "hover"))

```



## Correlations between QC metrics

This is a correlation matrix of the QC metrics considered so far which enables us to see whether lower quality samples fail across multiple metrics. 

``` {r correlation, fig.width = 10, fig.height= 10}
## create correlation plot 
corMergeStats<-cbind(mergeStats$Total.Sequences,mergeStats$total_deduplicated_percentage, mergeStats$Sequences.flagged.as.poor.quality,
                     mergeStats$PercentMTReads,
                     mergeStats$overall_alignment_rate,
                     mergeStats$DistinctReadPairs,
                     mergeStats$NRF,
                     mergeStats$PBC1,
                     propNucleosomesAll[,c(1,2)],
                     diptestStatsAll[,2],
                     decreasingProps[,1],
                     fripStats$BAMPeaks,
                     fripStats$FripMACS3PE)
colnames(corMergeStats)<- c('total\nreads', 'dedup', 'poor\nqual', 'mt\nreads', 'alignt\nrate', 'distinct\nreads', 
                            'NFR','PBC1',  "PropNFR", "PropMono", "DipP","Periodicity","NPeaks_PE", "FRIP_PE")
corMergeStats<- corMergeStats[, colSums(corMergeStats, na.rm = TRUE) != 0]
corrplot.mixed(cor(corMergeStats, use = "p"), order = 'AOE', tl.cex=0.9, tl.col = 'black')
```