---
title: '5hmC Data QC Report'
author: "Eilis Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: word_document
params:
  datapath: NA
---

```{r setenv, echo = FALSE, message = FALSE, warning = FALSE,results='hide'}

if(is.na(datapath)){
  stop("Filepath not specified on the command line")
}

library(ggplot2)
library(tidyverse)
library(qwraps2)


metaPath<-file.path(dataPath, "0_metadata")
peakQCPath<-file.path(dataPath, "4_calledPeaks", "QCOutput")
peakCountsPath<-file.path(dataPath, "5_peakCounts")

## load sample sheet
sampleSheet<-read.csv(file.path(metaPath, "sampleSheet.csv"))

## load frip statistics

fripFiles <- list.files(peakQCPath, pattern = "^FRIP")
seqSumStats <- do.call(rbind,lapply(file.path(peakQCPath, fripFiles),read.csv))

seqSumStats<-seqSumStats[match(seqSumStats$SampleName, sampleSheet$sampleID),]
seqSumStats$MACS2FRIP<-seqSumStats$PulldownReadsinMACS2Peaks/seqSumStats$BAMPulldownReads
seqSumStats$EPIC2FRIP<-seqSumStats$PulldownReadsinEPIC2Peaks/seqSumStats$BAMPulldownReads
seqSumStats$MACS2FRIPInput<-seqSumStats$ControlReadsinMACS2Peaks/seqSumStats$BAMPulldownReads
seqSumStats$EPIC2FRIPInput<-seqSumStats$ControlReadsinEPIC2Peaks/seqSumStats$BAMPulldownReads

seqSumStats <- seqSumStats %>% 
  mutate(BAMPulldownReads = BAMPulldownReads / 10^6) %>% 
  mutate(BAMControlReads = BAMControlReads / 10^6)

## load sex chromosome peak counts
sexCountsEpic<-read.table(file.path(peakCountsPath, "sexChr", "epic2.broadPeak"),header = TRUE)
sexCountsMacs<-read.table(file.path(peakCountsPath, "sexChr", "macs2.narrowPeak"),header = TRUE)
colnames(sexCountsEpic)<-unlist(lapply(strsplit(colnames(sexCountsEpic), "_S"), head, n = 1))
colnames(sexCountsMacs)<-unlist(lapply(strsplit(colnames(sexCountsMacs), "_S"), head, n = 1))

sexChrTotalsEpic<-t(aggregate(. ~ Chr, data = sexCountsEpic,sum))
sexChrTotalsMacs<-t(aggregate(. ~ Chr, data = sexCountsMacs,sum))
sexChrTotals<-data.frame("chrXepic" = as.numeric(sexChrTotalsEpic[-c(1:6), 1]), "chrYepic" = as.numeric(sexChrTotalsEpic[-c(1:6), 2]),"chrXmacs" = as.numeric(sexChrTotalsMacs[-c(1:6), 1]), "chrYmacs" = as.numeric(sexChrTotalsMacs[-c(1:6), 2]))
sexChrTotals$Sex<-sampleSheet$sex[match(gsub("[I|P]C", "", colnames(sexCountsEpic)[-c(1:6)]), make.names(sampleSheet$sampleID))]




## load gene body counts

"/gpfs/mrc0/projects/Research_Project-MRC190311/DNAhydroxySeq/mrcScz/5_peakCounts/GENCODE/geneCounts.summary"

```


This report summarises the data from CEGZ hydroxymethylation profiling, it implements a quality control pipeline adapted from https://doi.org/10.21203/rs.3.rs-667874/v1. 

## Data summary

Below are some summary statistics

```{r}
 our_summary1 <-
  list("Aligned reads from hydroxymethylome library" =
       list("mean (sd)" = ~ qwraps2::mean_sd(BAMPulldownReads, na.rm = TRUE)),
       "Aligned reads from input library"  =
       list("mean (sd)" = ~ qwraps2::mean_sd(BAMControlReads, na.rm = TRUE)),
       "MACS2 peaks (narrow)" =
       list("mean (sd)" = ~ qwraps2::mean_sd(MACS2Peaks, na.rm = TRUE)),
       "MACS2 FRIP" = 
       list("mean (sd)" = ~ qwraps::mean_sd(MACS2FRIP, na.rm = TRUE))
       "EPIC2 peaks (broad)" =
       list("mean (sd)" = ~ qwraps2::mean_sd(EPIC2Peaks, na.rm = TRUE)),

       )

whole <- summary_table(seqSumStats, our_summary1)
whole

```

```{r}
## violin plot of sequencing reads

seqSumStats %>% 
  select(BAMPulldownReads, BAMControlReads) %>% 
  gather(library, val) %>% 
  ggplot(aes(x=library, y=val)) + 
    geom_violin(trim=FALSE)+ 
  stat_summary(fun.data="mean_sdl", mult=1, 
                 geom="crossbar", width=0.2 ) +
  labs(
    caption = "Violin plot of number of aligned, deduplicated and filtered reads per sequencing library",
    tag = "Figure 1",
    x = "Library",
    y = "Aligned reads (millions)"
  ) + theme_bw() + geom_hline(yintercept=10)
```


```{r}
## scatter plot of sequencing reads

ggplot(seqSumStats, aes(x=BAMPulldownReads, y=BAMControlReads)) + 
  geom_point() + 
  labs(
    caption = "Scatterplot of number of aligned, deduplicated and filtered reads between matched sequencing libraries from the same sample",
    tag = "Figure 2",
    x = "Hydroxymethylome library (millions)",
    y = "Input library (millions)"
  ) + theme_bw() + geom_hline(yintercept=10) + geom_vline(xintercept=10)


```


```{r}
## violin plot of FRIP

seqSumStats %>% 
  select(MACS2FRIP, MACS2FRIPInput, EPIC2FRIP, EPIC2FRIPInput) %>% 
  gather(peaks, val) %>% 
  ggplot(aes(x=peaks, y=val)) + 
    geom_violin(trim=FALSE)+ 
  stat_summary(fun.data="mean_sdl", mult=1, 
                 geom="crossbar", width=0.2 ) +
  labs(
    caption = "Violin plot of fraction of reads in peaks",
    tag = "Figure 3",
    x = "",
    y = "Fraction of reads in peaks"
  ) + theme_bw() + geom_hline(yintercept=0.1)
```

To confirm sex 

```{r}

## check sex

ggplot(sexChrTotals, aes(x=chrXepic, y=chrYepic, color=Sex)) + 
  geom_point() + 
  labs(
    caption = "Scatterplot of reads in EPIC2 peaks on sex chromosomes",
    tag = "Figure 4",
    x = "X chr",
    y = "Y chr"
  ) + theme_bw() 

ggplot(sexChrTotals, aes(x=chrXmacs, y=chrYmacs, color=Sex)) + 
  geom_point() + 
  labs(
    caption = "Scatterplot of reads in MACS2 peaks on sex chromosomes",
    tag = "Figure 5",
    x = "X chr",
    y = "Y chr"
  ) + theme_bw() 
```

```{r}


```